In [62]:
import pandas as pd
import numpy as np
import random
import tensorflow 
from tensorflow import keras
from keras import Sequential
from keras.layers import Dense
import time

In [63]:
num_rows = 200
df = pd.DataFrame({
    "User ID": [random.randint(15000000, 16000000) for _ in range(num_rows)],
    "Gender": [random.choice(["Male", "Female"]) for _ in range(num_rows)],
    "Age": [random.randint(18, 60) for _ in range(num_rows)],
    "EstimatedSalary": [random.randint(15000, 100000) for _ in range(num_rows)],
    "Purchased": [random.choice([0, 1]) for _ in range(num_rows)],
})


In [64]:
df.head()

User ID  Gender  Age  EstimatedSalary  Purchased
0  15271008    Male   19            72164          1
1  15613290    Male   30            49652          0
2  15084158    Male   43            36924          0
3  15955135  Female   18            89787          0
4  15532467  Female   26            41304          1

In [65]:
df = df[['User ID','EstimatedSalary','Purchased']]

In [66]:
df.head()


User ID  EstimatedSalary  Purchased
0  15271008            72164          1
1  15613290            49652          0
2  15084158            36924          0
3  15955135            89787          0
4  15532467            41304          1

In [67]:
X = df.iloc[:,0:2]
y = df.iloc[:,-1]

In [68]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

In [69]:
from sklearn.model_selection import train_test_split

In [70]:
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size = 0.2,random_state =2)

In [71]:
X_train.shape

(160, 2)

In [72]:
 model = Sequential()
model.add(Dense(10,activation = 'relu', input_dim = 2))
model.add(Dense(10,activation = 'relu'))
model.add(Dense(1,activation = 'sigmoid'))

                
          

C:\Users\kanak\anaconda3\anaconda\Lib\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [73]:
model.summary()

Model: "sequential_3"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ dense_9 (Dense)                      │ (None, 10)                  │              30 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_10 (Dense)                     │ (None, 10)                  │             110 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_11 (Dense)                     │ (None, 1)                   │              11 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 151 (604.00 B)

 Trainable params: 151 (604.00 B)

 Non-trainable params: 0 (0.00 B)

In [74]:
model.compile(loss= 'binary_crossentropy',metrics = ['accuracy'])
start = time.time()

USing batch size we can change it to SGD or Batch gradient descent

In [75]:
#SGD  ---stochastic gradient descent
history  = model.fit(X_train,y_train,epochs= 10,batch_size =1)
print(time.time()-start)

Epoch 1/10
160/160 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.4735 - loss: 1029369.9375
Epoch 2/10
160/160 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.5513 - loss: 22792.7871
Epoch 3/10
160/160 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.5125 - loss: 25265.5039
Epoch 4/10
160/160 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6529 - loss: 19379.2617
Epoch 5/10
160/160 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.5585 - loss: 23285.1250
Epoch 6/10
160/160 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.5227 - loss: 24569.4512
Epoch 7/10
160/160 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.4514 - loss: 29614.3848
Epoch 8/10
160/160 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.4032 - loss: 27292.5117
Epoch 9/10
160/160 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.4215 - loss: 26899.6738
Epoch 10/10
160/160 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.5086 - loss: 22975.4043
5.033339977264404


IT TOOk 5.033 seconds to complete 10 epochs 

Now giving the batch size as the entire dataset

In [76]:
start_2 =time.time()
history_GD = model.fit(X_train,y_train,epochs =10, batch_size =160)
print(time.time()-start_2)

Epoch 1/10
1/1 ━━━━━━━━━━━━━━━━━━━━ 1s 551ms/step - accuracy: 0.4625 - loss: 5685.2944
Epoch 2/10
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 65ms/step - accuracy: 0.4812 - loss: 5925.3682
Epoch 3/10
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 72ms/step - accuracy: 0.4750 - loss: 6834.6670
Epoch 4/10
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 69ms/step - accuracy: 0.4875 - loss: 9825.5781
Epoch 5/10
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 72ms/step - accuracy: 0.5125 - loss: 16074.5371
Epoch 6/10
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 72ms/step - accuracy: 0.4875 - loss: 9272.0303
Epoch 7/10
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 70ms/step - accuracy: 0.5125 - loss: 17865.7852
Epoch 8/10
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 64ms/step - accuracy: 0.4625 - loss: 8638.8730
Epoch 9/10
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 115ms/step - accuracy: 0.5125 - loss: 18441.7090
Epoch 10/10
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 85ms/step - accuracy: 0.4875 - loss: 9233.8828
1.357173204421997


#FOR Batch GD it took only 1.35 ms